src/sources/stdin.rs

```diff
@@ -58,49 +62,106 @@ impl SourceConfig for StdinConfig {
+    // and have stdin sources and background thread synchronize on it.
+    //
+    // When source is builded it must:
+    // 1. enter critical section by locking mutex.
+    // 2. if sender isn't present start background thread and set sender.
+    // 3. gain receiver by calling subscribe on the sender.
+    // 4. release lock.
+    //
+    // When source has finished it should just drop it's receiver.
+    //
+    // Background thread should be started with copy of the sender.
+    //
+    // Once background thread want's to stop it must:
+    // 1. enter critical section by locking mutex.
+    // 2. if there are receivers it must abort the stop.
+    // 3. remove sender.
+    // 4. release lock.
+    //
+    // Although it's possible to implement this in a lock free, maybe even wait free manner,
+    // this should be easier to reason about and performance shouldn't suffer since this procedure
+    // is cold compared to the rest of the source.
+
+    let host_key = config
+        .host_key
+        .clone()
+        .unwrap_or(event::log_schema().host_key().to_string());
+    let hostname = hostname::get_hostname();
+
+    let mut guard = CRITICAL_SECTION
+        .lock()
+        .map_err(|_| "CRITICAL_SECTION poisoned.")?;
+    let receiver = if let Some(sender) = guard.as_ref() {
+        sender.subscribe()
+    } else {
+        let (sender, receiver) = channel(1024);
+        *guard = Some(sender.clone());
+
+        // Start the background thread
         thread::spawn(move || {
+            info!("Capturing STDIN.");
+
             for line in stdin.lines() {
                 match line {
                     Err(e) => {
                         error!(message = "Unable to read from source.", error = %e);
                         break;
                     }
-                    Ok(string_data) => {
-                        let msg = Bytes::from(string_data);
-                        while let Err(e) = tx.try_send(msg.clone()) {
-                            if e.is_full() {
-                                thread::sleep(Duration::from_millis(10));
+                    Ok(line) => {
+                        if sender.send(Bytes::from(line)).is_err() {
+                            // There are no active receivers.
+                            // Try to stop.
+                            let mut guard =
+                                CRITICAL_SECTION.lock().expect("CRITICAL_SECTION poisoned");
+                            if sender.receiver_count() > 0 {
+                                // A new new receiver has shown up.
+
+                                // It's fine not to resend the line since it came from
+                                // before this new receiver has shown up.
                                 continue;
                             }
-                            error!(message = "Unable to send event.", error = %e);
-                            break;
+                            guard.take();
```

Comment:
would be nice to make this an if/else

Reply:
Or even just negate the `if`, since that ends up being the fall through path.